### ISSUE TITLE
Ability to provide custom strcmp

### ISSUE THREAD

**Issue by alex-tee** [tlsa/libcyaml#164]
I'm writing extremely large files (100s of megabytes before compression) using cyaml and naturally it starts to get laggy when saving/loading large YAML files. having the ability to provide the custom strcmp function that libcyaml uses would allow me to use an optimized string interner to speed up string comparison for things that are repeated thousands of times in my YAML files like schema keys and flags

**Comment by alex-tee** [comment]
alternatively, having the option to use a string interner internally in cyaml would work as well (at the expense of higher RAM usage). cyaml could free it when it's done. I can try to implement this and send a patch if you think it's a good idea

**Comment by tlsa** [comment]
I like the idea of a client custom strcmp, it would allow the client to choose if it wants (Unicode) case insensitive matching.

**Comment by tlsa** [comment]
I'll look into it later today, thanks for the request!

**Comment by alex-tee** [comment]
awesome, thanks 👍

**Comment by tlsa** [comment]
Can you share a (compressed) copy of a typical 100MB YAML file and also the CYAML schema you have to load it?

**Comment by alex-tee** [comment]
My cyaml schemas are very complex sorry.. If it's not too much work I'd ask you to build zrythm: https://github.com/zrythm/zrythm . Most of the dependencies can be found in standard distro repositories and for the rest meson will attempt to fetch and build them if not found. Build instructions are here: https://github.com/zrythm/zrythm/blob/master/INSTALL.md

I uploaded a Zrythm project here <https://www.zrythm.org/downloads/Untitled%20Project%20(336).tar.xz>. You can pass it as a command line argument to `zrythm_launch --dummy <path to .zpj file>` and it will start loading the project via libcyaml. The .zpj file in this directory is a ZST-compressed YAML file created with libcyaml. Compressed it's around 1mb, uncompressed around 100mb. You can save the project with Ctrl-S and it will overwrite the file using cyaml. There is also a `--cyaml-log-level` option  that takes in one of: debug | info | warning | error and it will print the CYAML logs in the console. It's set to warning by default.

If that's too much hassle I can come up with some simpler code instead if you prefer in an hour or two. By the way, this is the interner I want to use: https://git.zrythm.org/zrythm/zrythm/src/branch/master/src/utils/symap.c#L47

**Comment by alex-tee** [comment]
actually I think the way this would work is cyaml would need to save an `int` or equivalent internally for each key/enum symbol, and instead of doing `strcmp` it would just check for int equality. I don't think just allowing to override `strcmp` will help in this case

**Comment by tlsa** [comment]
Please try this branch: [tlsa/fast-strings](https://github.com/tlsa/libcyaml/commits/tlsa/fast-strings).

I'd be very interested in any before/after performance stats you can measure with [tlsa/fast-strings](https://github.com/tlsa/libcyaml/commits/tlsa/fast-strings) and [main](https://github.com/tlsa/libcyaml/commits/main) branches.

**Comment by alex-tee** [comment]
it doesn't seem to make much difference:

```
100mb YAML file

-- original -O2 --
time to serialize: 2191ms
time to deserialize: 2416ms

-- fast-strings -O2 --
time to serialize: 2323ms
time to deserialize: 2510ms

-- original -O3 --
time to serialize: 2624ms
time to deserialize: 2406ms

-- fast-strings -O3 --
time to serialize: 2592ms
time to deserialize: 2392ms
```

I think the effort to calculate the string hash for the whole string and doing `strlen()` is as much work as just doing `strcmp()`

**Comment by tlsa** [comment]
> it doesn't seem to make much difference

Just to check, you have `ef96155f478c5dc589722b86a79b06e3058cc752`?

> I think the effort to calculate the string hash for the whole string and doing strlen() is as much work as just doing strcmp()

I hoped it would be quicker to get the string hash once and compare N hashes, than do N strcmp()s.

**Comment by alex-tee** [comment]
>Just to check, you have ef96155f478c5dc589722b86a79b06e3058cc752?

yeah although I merged the code to my meson-ified project https://git.zrythm.org/zrythm/zrythm-cyaml/src/branch/fast-strings

I'm making a test project for testing my CYAML schema (I managed to get something using `gcc -E`) will post it here in a bit to make testing easier. I'll also play around with the code a bit to see if I can come up with something

**Comment by tlsa** [comment]
I've just created 24c6262f2cc7e68ed93db37d632d56ee6509a0fb which optimises case insensitive string hashing.

**Comment by alex-tee** [comment]
I made this project to test more efficiently. it takes this file as input (extract first): https://www.zrythm.org/downloads/large-zrythm-project.yaml.tar.xz
https://git.zrythm.org/alex/cyaml-testing

All built with -O2 and tested using `build/large-yaml-test ~/Downloads/Untitled\ Project\ \(336\)/project.yaml`

24c6262
```
Time to deserialize: 3090ms
Time to serialize: 1867ms
```

ef96155
```
Time to deserialize: 3092ms
Time to serialize: 1868ms
```

master
```
Time to deserialize: 3031ms
Time to serialize: 1789ms
```

**Comment by alex-tee** [comment]
sorry my previous results were not -O2

24c6262
```
Average time to deserialize (5 tries, -O2): 2440ms
Average time to serialize (5 tries, -O2): 1727ms
```

ef96155
```
Average time to deserialize (5 tries, -O2): 2343ms
Average time to serialize (5 tries, -O2): 1734ms
```

master
```
Average time to deserialize (5 tries, -O2): 2463ms
Average time to serialize (5 tries, -O2): 1728ms
```

I also updated https://git.zrythm.org/alex/cyaml-testing/ to do 5 tries for each and display the optimization given

**Comment by danielealbano** [comment]
Why don't you use AVX2 for the string comparison? Not sure if it helps for this specific use case but the overhead introduced by having to invoke an external strcmp, as it is provided by glibc and can't be therefore inlined, is significant for short strings.

Here a simple implementation for short strings with related benchmark
https://quick-bench.com/q/seIszOXpaV6VPaj4D06eHiFQlbk

This specific implementation is for a project of mine where I can ensure that I have always 32 bytes of memory starting from the offset I want to compare, limiting then the comparison to the initial segment I care about passing a length, but a similar approach can be implemented or, in general, the AVX2 instructions for string comparison can be used directly instead of always invoking strcmp.

From the benchmark above, the overhead is quite significant, especially for strings that are longer than 15 chars

__attribute__((__target__("avx2")))` is only for the quick-bench website as it's not possible to specify which compiler optimization have to be enabled, in the real world would be better to use an intrinsic to check if the cpu supports AVX2.

AVX2 is pretty common nowadays so unless you need to support 10 years old CPUs it's a very good alternative, at least on Intel/AMD CPUs, I don't have any code for the NEON (for now).

**Comment by alex-tee** [comment]
@danielealbano isn't the implementation of strcmp in glibc already axv-optimized? I believe strcmp() in glibc automatically figures out which version of strcmp to use based on your cpu. if you profile it it shows something like `--strcmp_avx2()`. here is a screenshot from some of my profiling results with kcachegrind

![Screenshot from 2021-08-15 00-20-16](https://user-images.githubusercontent.com/3407655/129462491-3e816a6f-9d3d-4057-a3ff-b41a69ed7008.png)

**Comment by alex-tee** [comment]
by the way, I just realized that string comparisons are only done during load. my biggest concern is the time it takes to save, since my software does auto-save every minute or so. loading time is not a concern to me since it only happens on startup in my software - although optimization would still be welcome

**Comment by danielealbano** [comment]
> @danielealbano isn't the implementation of strcmp in glibc already axv-optimized? I believe strcmp() in glibc automatically figures out which version of strcmp to use based on your cpu. if you profile it it shows something like `__strcmp_avx2()`. here is a screenshot from some of my profiling results with kcachegrind
> 
> ![Screenshot from 2021-08-15 00-20-16](https://user-images.githubusercontent.com/3407655/129462491-3e816a6f-9d3d-4057-a3ff-b41a69ed7008.png)

Yes it is, infact I was referring to the overhead brought by having to call the external strcmp implemented in glibc that, as you can see from the benchmark, goes from about 6 to further more.

I'm case of having to compare thousands or tens of thousands of strings using an avx2 implementation that can be therefore inlined by the compiler providers much better performance.

Of course, being your use case focused on saving and not loading, there is little point in taking advantage of this optimization.

I wonder, why not saving in the background in a separate thread?
Or, even better, can't you split your single yaml in a number of them so you can save them in parallel and then "compress" them in an archive like format? It's a fairly common practice.

**Comment by alex-tee** [comment]
> I wonder, why not saving in the background in a separate thread?

a lot of things in my yaml relate to other things in the same yaml so the user should not be able to change anything that would be serialized while the yaml is being saved. we've had issues with corrupted projects before when I was saving in the background

>Or, even better, can't you split your single yaml in a number of them so you can save them in parallel 

this is a good idea, thanks. perhaps instead of a single project file I could split it into various files that make up the project and save them in parallel.

>and then "compress" them in an archive like format? It's a fairly common practice.

I already do this with ZST, but the time overhead to save is mostly due to cyaml serializing the whole data into the yaml first. a user reported waiting for 20 seconds during save so this is why I'm trying to optimize it. size is not an issue thanks to compression

**Comment by tlsa** [comment]
> I also updated https://git.zrythm.org/alex/cyaml-testing/ to do 5 tries for each and display the optimization given. if you want to use that for quick testing you can pass `--force-fallback-for=libcyaml` to meson and then edit the code inside `subprojects/libcyaml` directly then you can just rebuild with `ninja -C build`

Thanks for this test case. It looks like string comparison is not the bottleneck at all.  This explains why changing the way it works had no real impact on the timings.

For saving there is no need for string comparison in LibCYAML anyway. Most of the time is spent in the `libyaml` event emitter and event initialisation code.  There's no way to optimise this in LibCYAML short of writing my own low-level YAML write implementation and avoiding `libyaml`.

For loading, the string comparison is used to look up whatever has been read from the YAML input in the schema.  However, most of the time is actually spent in `libyaml` code (`yaml_parser_parse`). Again, there's not much scope for optimising it in LibCYAML.

@danielealbano's idea of splitting into separate files is a good one.

Another idea might be to deep clone your data structure on the main thread (which should be relatively quick), and spawning a thread for writing the cloned data out to file, so it doesn't block anything.

**Comment by alex-tee** [comment]
sounds like there's not too much that can be done in cyaml optimization-wise. feel free to close this unless custom strcmp makes sense for other things

>Another idea might be to deep clone your data structure on the main thread (which should be relatively quick), and spawning a thread for writing the cloned data out to file, so it doesn't block anything.

deep cloning on the main thread sounds like it will work since most if not all of my (serializable) data writes are done on the main thread. this is likely what I'll end up doing as I prefer to have all the data in one file

thanks for taking the time!
